addons/payment_ingenico/models/payment.py

```diff
@@ -151,8 +151,14 @@ def filter_key(key):
         items = sorted((k.upper(), v) for k, v in values.items())
         sign = ''.join('%s=%s%s' % (k, v, key) for k, v in items if v and filter_key(k))
         sign = sign.encode("utf-8")
-        shasign = sha1(sign).hexdigest()
-        return shasign
+
+        hash_function = self.env['ir.config_parameter'].sudo().get_param('payment_ingenico.hash_function', 'sha1').lower()
+        if hash_function not in ['sha1', 'sha256', 'sha512']:
```

Comment:
This is really nitpicking at this point but I too was bugged for a second reading this. And with `hashlib.new` not caring about the case it might be clearer to write this as follows:
```suggestion
        hash_function = self.env['ir.config_parameter'].sudo().get_param('payment_ingenico.hash_function')
        if not hash_function or hash_function.lower() not in ['sha1', 'sha256', 'sha512']:
```
It's 100% a matter of taste so you get to pick whichever you want ;)